from django import forms
from django.core.exceptions import ValidationError

from .functions import *


class AffiliationForm(forms.ModelForm):
    classes_list = get_classes_list()
    races_list = get_races_list()

    class Meta:
        model = Character
        fields = ['race', 'character_class']


class InformationForm(forms.Form):
    backgrounds_list = get_backgrounds_list()
    alignment_list = get_alignment_list()

    name = forms.CharField(label="Имя персонажа")
    background = forms.ChoiceField(label='Предыстория', choices=backgrounds_list)
    alignment = forms.ChoiceField(label='Мировоззрение', choices=alignment_list)
    player_name = forms.CharField(label="Имя игрока")
    level = forms.IntegerField(label='Уровень')


class PurchaseForm(forms.Form):
    strength = forms.IntegerField(min_value=8, max_value=15, initial=8)
    dexterity = forms.IntegerField(min_value=8, max_value=15, initial=8)
    physique = forms.IntegerField(min_value=8, max_value=15, initial=8)
    intelligence = forms.IntegerField(min_value=8, max_value=15, initial=8)
    wisdom = forms.IntegerField(min_value=8, max_value=15, initial=8)
    charisma = forms.IntegerField(min_value=8, max_value=15, initial=8)

    def clean(self):
        cleaned_data = super().clean()

        field_names = ['strength', 'dexterity', 'physique', 'intelligence', 'wisdom', 'charisma']

        used_points = sum(cleaned_data.get(field, 8) or 8 for field in field_names)

        if used_points > 75:
            raise ValidationError("Вы можете использовать только 27 очков для закупки характеристик.")

        return cleaned_data
